File: tunnels.py
from collections import deque



n, m, k = map(int, input().split())

tunnels = []
for i in range(n-1):
    tunnels.append([int(x) for x in input().split()])

racers = []
for i in range(m):
    racers.append([int(x) for x in input().split()])

end_checkpoint = int(input())
c = int(input())
specials = []
for i in range(c):
    specials.append(int(input()))




all_dist = []
for i in range(n):
    dist =[0 for _ in range(n)]
    all_dist.append(dist)


for i in range(1,n+1):
    for tunnel in tunnels: 
        if i in tunnel: 
            other = tunnel[0] if tunnel[0]!=i else tunnel[1]
            all_dist[i-1][other-1] = 1
            all_dist[other-1][i-1] = 1



print(all_dist)

dist = [-1] * n

q = deque()

dist[end_checkpoint-1] = 0
q.append(end_checkpoint-1)

# BFS
while q:
    u = q.popleft()
    for v in range(n):
        if all_dist[u][v] == 1 and dist[v] == -1:
            dist[v] = dist[u] + 1
            q.append(v)

        
for i in range(n):
    print(f"dist[{i+1}] = {dist[i]}")



sorted(specials, key=lambda x: dist[x], reverse=True)
passed_special = [0 for _ in specials]

can_pass = [[True for _ in range(len(specials))] for _ in range(len(racers))]
vel_specials = [[0 for _ in range(n)] for _ in specials]
print(specials)

speeds = []
for racer in racers: 
    speeds.append(1/(racer[1]))
passing_per_special = [] 

for special in specials:
    arrivals = []

    for r_index, racer in enumerate(racers):
        start = racer[0]
        speed = 1/racer[1]

        if dist[special-1] < dist[start-1]:
            arrivals.append((speed, r_index)) 

    
    arrivals.sort(reverse=True)

    passing = [r_index for  _, r_index in arrivals[:k]]

    passing_per_special.append(passing)

print("special")
print(passing_per_special)

#per ogni special ho una lista di velocità che i giocatori hanno a quel punto 
times = []

for j, racer in enumerate(racers):
    eliminated = False
    print(j)
    for i, special in enumerate(specials):
        if dist[special-1] < dist[racer[0]-1]: 
            if j not in passing_per_special[i]:
                eliminated = True
    if eliminated:
        times.append(-1)
    else:
        times.append(racer[1] * dist[racer[0]-1])




print(times)

File: tunnels.cpp
#include <iostream>
#include <vector>
#include <deque>
#include <algorithm>
#include <queue>

using namespace std;

int main()
{
    int n, m, k;
    cin >> n >> m >> k;

    vector<vector<int>> tunnels(n - 1, vector<int>(2));
    for (int i = 0; i < n - 1; i++)
    {
        cin >> tunnels[i][0] >> tunnels[i][1];
    }

    vector<vector<int>> racers(m, vector<int>(2));
    for (int i = 0; i < m; i++)
    {
        cin >> racers[i][0] >> racers[i][1];
    }

    int end_checkpoint;
    cin >> end_checkpoint;

    int c;
    cin >> c;

    vector<int> specials(c);
    for (int i = 0; i < c; i++)
    {
        cin >> specials[i];
    }

    vector<vector<int>> adj(n); 

    for (auto &t : tunnels)
    {
        int u = t[0] - 1;
        int v = t[1] - 1;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    vector<int> dist(n, -1);
    queue<int> q;
    dist[end_checkpoint - 1] = 0;
    q.push(end_checkpoint - 1);

    while (!q.empty())
    {
        int u = q.front();
        q.pop();
        for (int v : adj[u])
        { // solo vicini
            if (dist[v] == -1)
            {
                dist[v] = dist[u] + 1;
                q.push(v);
            }
        }
    }

    sort(specials.begin(), specials.end(), [&](int x, int y)
         { return dist[x] > dist[y]; });

    vector<int> passed_special(specials.size(), 0);

    vector<double> speeds;
    for (auto &racer : racers)
    {
        speeds.push_back(1.0 / racer[1]);
    }

    vector<vector<int>> passing_per_special;

    for (auto special : specials)
    {
        vector<pair<double, int>> arrivals;

        for (int r_index = 0; r_index < m; r_index++)
        {
            int start = racers[r_index][0];
            double speed = 1.0 / racers[r_index][1];

            if (dist[special - 1] < dist[start - 1])
            {
                arrivals.emplace_back(speed, r_index);
            }
        }

        sort(arrivals.rbegin(), arrivals.rend());

        vector<int> passing;
        for (int i = 0; i < (int)arrivals.size() && i < k; i++)
        {
            passing.push_back(arrivals[i].second);
        }

        passing_per_special.push_back(passing);
    }

    vector<int> times;

    for (int j = 0; j < m; j++)
    {
        bool eliminated = false;
        for (int i = 0; i < (int)specials.size(); i++)
        {
            if (dist[specials[i] - 1] < dist[racers[j][0] - 1])
            {
                if (find(passing_per_special[i].begin(), passing_per_special[i].end(), j) == passing_per_special[i].end())
                {
                    eliminated = true;
                    break;
                }
            }
        }
        if (eliminated)
        {
            times.push_back(-1);
        }
        else
        {
            times.push_back(racers[j][1] * dist[racers[j][0] - 1]);
        }
    }

    cout << "[";
    for (size_t i = 0; i < times.size(); i++)
    {
        cout << times[i];
        if (i != times.size() - 1)
            cout << ", ";
    }
    cout << "]" << endl;

    return 0;
}